Add an information_schema.role_grants table

lhofhansl opened the pull request:
Update:
this is now called information_schema.role_authorization_descriptors
The schema of this table is:
| role_name | grantor | grantor_type | grantee | grantee_type | is_grantable |

Connectors can use that to make role grant information available to (admin) users by implementing ConnectorMetadata.listAllRoleGrants, with an option to filter by predicates for optimization.

Currently this is implemented in the Hive Connector only.

grantor and grantor_type are not yet supported and always null. (but it could be added in a separate DR).

Continued from #3232

lhofhansl commented:
@kokosing FYI

Note that I have not done the predicate pushdown, yet. This has most of the plumbing. The caching seems to be fairly effective already, though.

```
presto> select count(*) from information_schema.role_grants,information_schema.role_grants,information_schema.role_grants,information_schema.role_grants,information_schema.role_grants,information_schema.role_grants,information_schema.role_grants,information_schema.role_grants,information_schema.role_grants,information_schema.role_grants,information_schema.role_grants;
   _col0   
-----------
 362797056 
(1 row)

Query 20200424_023942_00015_7issh, FINISHED, 1 node
Splits: 198 total, 198 done (100.00%)
0:00 [126 rows, 40B] [477 rows/s, 152B/s]
```

Although I assume there's some other optimization at work. :)

Is there a place where the information_schema is documented? (I could not find any)

lhofhansl commented:
For the predicate push down I can think of three scenarios:
1. point values for role_name - `WHERE role_name = 'x'`, or `WHERE role_name in (...)`, etc.
2. arbitrary predicate on role_name - `WHERE role_name LIKE ...`, etc.
3a. point values for grantee and grantee_type - `WHERE grantee = 'x' AND grantee_type = 'USER'`, or `WHERE (grantee, grantee_type) IN (('x', 'USER', 'y', 'ROLE', ...)`
3b. point values for grantee (name) only.

For (1) I'd loop over the passed values (up to a limit) and call `metastore.listPrincipals`
For (2) I'd retrieve all roles with `metadata.listRoles`, then filter them by the predicate and retrieve the remaining ones with `metadata.listPrinicipals`. That way any predicate would work. Perhaps (1) should just do the same.
For (3a) this looks quite tricky to me.
For (3b) I'd loop over the passed grantee name and retrieve them with two calls `metadata.listRoleGrants` one for USER and one for ROLE. This is still a pretty significant optimization when you look for the roles granted to a specific user.

lhofhansl commented:
Added predicate pushdown as described above.

For role names it supports all kinds of predicates (since we can enumerate all existing roles and then filter them), grantee names support "point values" only. To avoid surprises the grantee_type does not have to the specified, the code will retrieve both USERs and ROLEs of that name.

The "public" role is not listed, since that mapping is not actually stored anywhere and so cannot be retrieved when querying through roles.

There's a basic test. If there's a good place to test the predicate pushdown, please point me to it.

Other than perhaps more test and a place to put the documentation, this should be good for review.

@kokosing @martint

lhofhansl commented:
Looking at the test-failures...

lhofhansl commented:
All tests fixed now. @kokosing

lhofhansl commented:
Expanded the test a bit to include a predicate filter.

kokosing commented on presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java:
```diff
@@ -2341,6 +2341,12 @@ public void dropRole(ConnectorSession session, String role)
     }
 
+    @Override
+    public Set<RoleGrant> listPrincipals(ConnectorSession session, String role)
```
`listGrantedPrincipals`?

lhofhansl replied:
Yeah, that's better.

kokosing commented on presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/HiveMetastore.java:
```diff
@@ -105,6 +108,11 @@
 
     void revokeRoles(Set<String> roles, Set<HivePrincipal> grantees, boolean adminOption, HivePrincipal grantor);
 
+    default Set<RoleGrant> listPrincipals(String role)
```
remove `default`? Other methods do not have `default` here?

lhofhansl replied:
New update soon.

kokosing commented on presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java:
```diff
@@ -4962,6 +4962,74 @@ public void testShowColumnMetadata()
+    @Test
+    public void testRoleGrants()
+    {
+        assertUpdate("CREATE ROLE test_role_grant1");
```
Can you test different users here? Like admin who sees everything, user that sees only his role grants and user that does not see anything because has no role grants?

kokosing commented on presto-tests/src/test/java/io/prestosql/tests/TestInformationSchemaConnector.java:
```diff
@@ -52,12 +52,12 @@
     public void testBasic()
```
can you please add tests for `role_grants` here as well?

lhofhansl replied:
There are 0 tests for any other role or access related stuff in there. I assume it's because it does not work in all connectors.

kokosing replied:
> it's because it does not work in all connectors

This test is using dedicated testing connector called `CountingMockConnector`

> There are 0 tests for any other role or access related stuff

That is a mistake (we didn't catch that). I think having such tests bring a lot of value. Could you please add them? Adding tests for other tables could be done in separate commits.

lhofhansl replied:
I added some tests....

kokosing commented on presto-tests/src/test/java/io/prestosql/tests/TestInformationSchemaConnector.java:
```diff
@@ -52,12 +52,12 @@
         assertQuery("SELECT count(*) FROM tpch.information_schema.schemata", "VALUES 10");
-        assertQuery("SELECT count(*) FROM tpch.information_schema.tables", "VALUES 80");
-        assertQuery("SELECT count(*) FROM tpch.information_schema.columns", "VALUES 583");
+        assertQuery("SELECT count(*) FROM tpch.information_schema.tables", "VALUES 81");
```
why this numbers got higher?

lhofhansl replied:
This is because these include the new tables and columns in `information_schema` so both tables and columns have more values.

kokosing commented on presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java:
```diff
@@ -2341,6 +2341,12 @@ public void dropRole(ConnectorSession session, String role)
         return accessControlMetadata.listRoles(session);
```
@electrum Can you please take a look at this pull request too? I would like to have a second pair of eyes before adding any new table to `information_schema`?

kokosing commented:
I like the idea of system table, I regret that I haven't thought about this in the first place. However, I am not sure if it is easy to implement predicate pushdown for that easily.

lhofhansl commented:
There's certainly precedence for non-standard information_schemas:

- https://dev.mysql.com/doc/refman/8.0/en/information-schema.html
- https://www.postgresql.org/docs/9.1/information-schema.html
- https://cloud.google.com/bigquery/docs/information-schema-intro
- https://docs.snowflake.com/en/sql-reference/info-schema.html
- https://docs.oracle.com/cd/E19078-01/mysql/mysql-refman-5.0/information-schema.html

Without predicate pushdown I do not think that this would be useful.

lhofhansl commented:
Another question... So if had a Hive catalog called `x`. There would then be a table `x.system.role_grants`, right?

What if somebody has a `system` schema already in their Hive catalog? How would we ensure that there would be no name collisions?

lhofhansl commented:
Ok. I found a way for doing this, including some more limited predicate pushdown (limit is not supported and only single values).

Before I go on with this, let's please agree to do so and what the table should be called exactly. :)

I still think it's much better to expose this along `ROLES`, `ENABLED_ROLES`, and `APPLICABLE_ROLES` in `information_schema`

electrum commented:
I think we’d need to reserve `system` as a special name for Presto (in the Hive connector). We could make that name configurable if it’s a problem.

My main concern is that the implementation in the engine is specific to Hive. I’m not familiar enough with roles in the SQL standard to know what this should look like and will need to research it. Maybe @martint knows.

lhofhansl commented:
Thanks @electrum . I have a change more or less ready for the system table approach (was actually a good opportunity to understand the bootstrapping process better). The only problem left is *not* to install that table when SqlStandardAccessControl is not enabled for a catalog.

For now I'll hold off until I hear more.

lhofhansl commented:
For the `information_schema` we have more connectors that do support standard SQL access control: PostgreSQL, Phoenix, Oracle, and others.

In the future I'd envision that we could expose ROLES, ENABLED_ROLES, APPLICABLE_ROLES, TABLE_PRIVILEGES (and hopefully ROLE_GRANTS) for those as well.

I suppose the general question is how much of that we allow to do via Presto and for how much of these we'd defer to the underlying databases.

@kokosing @electrum @martint

lhofhansl commented:
I started a slack discussion on this.

I did find this in the standard (https://crate.io/docs/sql-99/en/latest/chapters/16.html#the-information-schema):

`The total number of Views in INFORMATION_SCHEMA, and their exact definition, is non-standard because the SQL Standard allows implementors to add additional Views, as well as to add additional Columns to the Standard- defined Views, to describe additional, implementation-defined features. However, except for one exception, the View descriptions that follow must all be supported by your SQL DBMS.`

And this (http://www.contrib.andrew.cmu.edu/~shadow/sql/sql1992.txt):

`         An implementation may define objects that are associated with
         INFORMATION_SCHEMA that are not defined in this Clause. An im-
         plementation may also add columns to tables that are defined in
         this Clause.`

So added a table to the i`information_schema` really seems to be the right way here.

lhofhansl commented:
Based on discussions on slack I propose the following: 

Add `information_schema.ROLE_AUTHORIZATION_DESCRIPTORS` with the following shape:
(role, grantor, grantor_type, grantee, grantee_type, is_grantable)

Note: Technically the existong ROLES table and this table should be in a schema called `definition_schema`.
I'm proposing - just like ROLES - to put it in `information_schema`, but again we wouldn't be SQL standard compliant. (Perhaps with a followup PR to move both into a schema called `definition_schema`.)

lhofhansl commented:
I'll post an update soon that adds the `definition_schema` with both `roles` and `role_authorization_descriptors` in it. `roles` will also remain available `information_schema` for backwards compatibility.

In order to avoid a large amount of code both schemas are served from the current InformationSchema* classes. If that is not acceptable I can disentangle that too.

lhofhansl commented:
Push as a new commit, so that we can see the difference. If OK, I'll squash before commit.

@martint , @electrum , @kokosing please have a look.

lhofhansl commented:
```
presto> show schemas;
       Schema       
--------------------
[...]
 definition_schema  
 information_schema 
[...]
(11 rows)

Query 20200501_020552_00067_7ij2u, FINISHED, 1 node
Splits: 19 total, 19 done (100.00%)
0:00 [11 rows, 129B] [43 rows/s, 514B/s]

presto> show tables in definition_schema;
             Table              
--------------------------------
 role_authorization_descriptors 
 roles                          
(2 rows)

Query 20200501_020609_00068_7ij2u, FINISHED, 1 node
Splits: 19 total, 19 done (100.00%)
0:00 [2 rows, 89B] [6 rows/s, 307B/s]

presto> show tables in information_schema;          
      Table       
------------------
 applicable_roles 
 columns          
 enabled_roles    
 roles            
 schemata         
 table_privileges 
 tables           
 views            
(8 rows)

Query 20200501_020616_00069_7ij2u, FINISHED, 1 node
Splits: 19 total, 19 done (100.00%)
0:00 [8 rows, 300B] [32 rows/s, 1.19KB/s]

presto> select * from definition_schema.role_authorization_descriptors;
 role_name | grantor | grantor_type | grantee | grantee_type | is_grantable 
-----------+---------+--------------+---------+--------------+--------------
 admin     | NULL    | NULL         | admin   | USER         | YES          
 admin     | NULL    | NULL         | lars    | USER         | NO           
 admin     | NULL    | NULL         | user    | USER         | YES          
 admin     | NULL    | NULL         | test    | USER         | NO           
 auth      | NULL    | NULL         | lars    | USER         | NO           
 auth      | NULL    | NULL         | auth    | USER         | NO           
 test      | NULL    | NULL         | auth    | ROLE         | NO           
 test      | NULL    | NULL         | lars    | USER         | YES          
(8 rows)

Query 20200501_020635_00070_7ij2u, FINISHED, 1 node
Splits: 17 total, 17 done (100.00%)
0:00 [8 rows, 360B] [29 rows/s, 1.28KB/s]
```

lhofhansl commented:
Fixing some tests.

lhofhansl commented:
I'll look at the remaining tests when we have agreed on final way to expose this information.

I have no explored four different ways:
1. `SHOW PRINCIPALS role` and `SHOW ROLE GRANTS FOR`
2. A new table in `information_schema` (I called it `role_grants`, but we can just name it `role_authorization_descriptors` and add the `grantor` columns)
3. A `system` table of the same shape.
4. `roles`, and `role_authorization_descriptors` in a new `definition_schema`. Also keep a `roles` table in `information_schema` for backward compatibility.

All of these are good options (IMHO). _Let's pick one._ :)
(I'd vote for (2) followed by (4), followed by (3) and (1)... In that order.)

kokosing commented:
I would go with 2 and possibly with 4 later.

We need to to get conclusion before we move any further. @martint @electrum What is your voice here?

lhofhansl commented:
Rebased yesterday evening.

Yep... (2) or (4) seem to be the best options.
(4) is a bit more SQL compliant, but (2) is more in line with what we have already in Presto, a smaller less disruptive change, and perhaps less of a surprise to users (no new schema).

If we decide for (2) I'll remove the last commit and continue from there, for (4) I'd squash the two commits and then go from there.

This is also a friendly (ping) :)
(I'd like to do an internal release with this functionality and would prefer to do it with the open source implementation.)

lhofhansl commented:
@martint @electrum ping... Just point me in the right direction and I'll continue.

lhofhansl commented:
Rebased

@kokosing @martint @electrum

lhofhansl commented:
Any updates?

Perhaps let me just pick an implementation (option 2 from above) to use at Salesforce for now.
(but I'd prefer not to do a fork even if it's temporary)

electrum commented:
I didn't realize that the existing role tables in `information_schema` are non-standard. That was probably a miss when we were reviewing and merging that feature. But since we already have those, I think it makes sense to be consistent with those. Let's go with option 2.

I'm still concerned that the comments in `InformationSchemaPageSource` talk about the Hive metastore. How specific to Hive is this implementation? Would it be applicable to other connectors?

lhofhansl commented:
Thanks @electrum . That's an excellent point. These are Hive Metastore limitations.

There are two options, I think.
1. Say that `listRoleGrants` and `listPrincipals` is the lowest common denominator. Then all connectors would have to implement those two.
2. Think of a more general interface in Metadata, and connectors are free to implement that as they see fit. The trick would be to still benefit from predicate push down.

OK... I'll go with (2) above + change the name of the table to `role_authorization_descriptors`.
And I'll think through a metadata interface that is better applicable to all connectors.

lhofhansl commented:
OK... Updated to use `information_schema.role_authorization_descriptors`.
The syntax should be final.

I will now look into how to add a more useful method to `Metadata` and `ConnectorMetadata` to allow connectors retrieving all role grants. Something like 'getAllRoleGrants` with optional role and grantees passes for filtering.

lhofhansl commented:
Good. Functional tests. That's a baseline. I'll have a change to push the Hive specific logic into the Hive Connector.

lhofhansl commented:
OK... Have a look. The Hive logic is now where it should be: `io.prestosql.plugin.hive.security.SqlStandardAccessControlMetadata` in a new method: `listAllRoleGrants` that is being passed the relevant predicated to optionally filter the results.

lhofhansl commented on presto-hive/src/main/java/io/prestosql/plugin/hive/security/SqlStandardAccessControlMetadata.java:
```diff
@@ -84,6 +87,62 @@ private static void checkRoleIsNotReserved(String role)
+        long count = 0;
+        if (grantees.isPresent()) {
+            TOP:
+            for (String grantee : grantees.get()) {
```
If there's a more elegant to phrase this let me know.
Instead of the `break label` that could build() and return as well.

kokosing replied:
extract a method and use `return`?

kokosing commented on presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/file/FileHiveMetastore.java:
```diff
@@ -814,6 +814,16 @@ public synchronized void revokeRoles(Set<String> roles, Set<HivePrincipal> grant
+    @Override
+    public synchronized Set<RoleGrant> listGrantedPrincipals(String role)
+    {
+        ImmutableSet.Builder<RoleGrant> result = ImmutableSet.builder();
+        result.addAll(listRoleGrantsSanitized().stream()
+                .filter(grant -> grant.getRoleName().equals(role))
+                .collect(toSet()));
+        return result.build();
```
```suggestion
        return listRoleGrantsSanitized().stream()
                .filter(grant -> grant.getRoleName().equals(role))
                .collect(toImmutableSet());
```

lhofhansl replied:
I was going by `listRoleGrants` below. :)  I'll change them both.

lhofhansl replied:
Done

kokosing commented on presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/ThriftHiveMetastoreClient.java:
```diff
@@ -418,6 +420,15 @@ private void removeGrant(String role, String granteeName, PrincipalType granteeT
+            throws TException
+    {
+        GetPrincipalsInRoleRequest request = new GetPrincipalsInRoleRequest(role);
+        GetPrincipalsInRoleResponse resp = client.get_principals_in_role(request);
```
s/resp/response

lhofhansl replied:
Done

kokosing commented on presto-hive/src/main/java/io/prestosql/plugin/hive/security/SqlStandardAccessControl.java:
```diff
@@ -95,6 +95,7 @@
     public static final String ADMIN_ROLE_NAME = "admin";
     private static final String INFORMATION_SCHEMA_NAME = "information_schema";
     private static final SchemaTableName ROLES = new SchemaTableName(INFORMATION_SCHEMA_NAME, "roles");
+    private static final SchemaTableName ROLE_GRANTS = new SchemaTableName(INFORMATION_SCHEMA_NAME, "role_authorization_descriptors");
```
rename this variable to match the value it holds

lhofhansl replied:
Done

kokosing commented on presto-hive/src/main/java/io/prestosql/plugin/hive/security/SqlStandardAccessControlMetadata.java:
```diff
@@ -84,6 +87,62 @@ private static void checkRoleIsNotReserved(String role)
+                            break TOP;
+                        }
+                        // Filter out the "public" role since it is not explicitly granted in Hive.
+                        if ("public".equals(grant.getRoleName())) {
```
use some constant for `"public"`. There should be some already.

lhofhansl replied:
The trick is that we absolutely need to honor the limit the first time it hits. Each loop here is a roundtrip to the metastore (that's why the logic is a bit convoluted with the break label).

I could pass the builder to a method and that method returns a boolean to indicate whether we had hit the limit or not (so that the outer loops knows it's done).

Not sure that is much prettier. At least with the behavior is explicit. Lemme mull over that a bit more.

lhofhansl replied:
On "public". Honestly, none that I could find. The "public" string is all over the place in the Hive connector. There're private constants iniThriftMetastoreUtil, I could make those public.

lhofhansl replied:
Added a constant to this class at least.

lhofhansl replied:
Cleaned up the logic too, extracted methods, etc.

kokosing commented on presto-hive/src/main/java/io/prestosql/plugin/hive/security/SqlStandardAccessControlMetadata.java:
```diff
@@ -84,6 +87,62 @@ private static void checkRoleIsNotReserved(String role)
+        }
+
+        // if no roles or grantees were passed get a list of roles from the metastore
+        for (String role : roles.orElseGet(() -> metastore.listRoles())) {
```
please extract one method for collecting output for `grantees` and one method for `roles`

lhofhansl replied:
Done.

kokosing commented on presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java:
```diff
@@ -4973,6 +4973,87 @@ public void testShowColumnMetadata()
     }
 
+    @Test
+    public void testRoleGrants()
```
update test name to match `role_authorization_descriptors`

lhofhansl replied:
Done

kokosing commented on presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java:
```diff
@@ -4973,6 +4973,87 @@ public void testShowColumnMetadata()
+        assertQuery(
+                "SELECT * FROM information_schema.role_authorization_descriptors WHERE grantee_type = 'ROLE'",
+                "VALUES ('test_role_grant2', null, null, 'test_role_grant1', 'ROLE', 'NO')");
+
```
Please add some test with `LIMIT`

lhofhansl replied:
Done

kokosing commented on presto-main/src/main/java/io/prestosql/connector/informationschema/InformationSchemaPageSource.java:
```diff
@@ -346,13 +355,35 @@ private void addRolesRecords()
+        for (RoleGrant grant : metadata.listAllRoleGrants(session, catalogName, roles, grantees, limit)) {
+            addRecord(
+                    grant.getRoleName(),
+                    null, // grantor
```
It would be nice to have these defined in `RoleGrant`, but being unused currently. So other connector implementations could benefit from that.

lhofhansl replied:
Let's do that in a separate PR if that's ok.

kokosing commented on presto-main/src/main/java/io/prestosql/connector/informationschema/InformationSchemaPageSource.java:
```diff
@@ -346,13 +355,35 @@ private void addRolesRecords()
+    private void addRoleAuthorizationDescriptorRecords()
+    {
+        try {
+            accessControl.checkCanShowRoles(session.toSecurityContext(), catalogName);
```
I think we need dedicated access control method for that.

lhofhansl replied:
I was thinking about that. Personally I thought there's access to roles or not.

One could argue that access to ROLE_AUTHORIZATION_DESCRIPTORS is a superset access right (I can get the list of roles from that table as well.) 
Separate access rights seems a bit overkill - happy to put it in if you feel strongly, though.

kokosing replied:
Presto SPI do not assume much about authorization system implementation. Having separate method allows it to decide how it should behave. I would prefer to have separate method.

lhofhansl replied:
OK... Added that.

kokosing commented on presto-main/src/main/java/io/prestosql/connector/informationschema/InformationSchemaMetadata.java:
```diff
@@ -191,19 +195,24 @@ public ConnectorTableProperties getTableProperties(ConnectorSession session, Con
-        if (prefixes.equals(table.getPrefixes())) {
-            return Optional.empty();
+        Set<QualifiedTablePrefix> prefixes = table.getPrefixes();
+        if (isTablesEnumeratingTable(table.getTable()) && table.getPrefixes().equals(defaultPrefixes(catalogName))) {
```
I don't think we should do such ifs. I would prefer to try to update `prefixes` and `grantees` always. And if any of them is not changed return `Optional.empty()` (as it was before).

lhofhansl replied:
Agreed. Much better.

lhofhansl replied:
Done.

kokosing commented on presto-main/src/main/java/io/prestosql/connector/informationschema/InformationSchemaMetadata.java:
```diff
@@ -182,7 +186,7 @@ public ConnectorTableProperties getTableProperties(ConnectorSession session, Con
         }
 
         return Optional.of(new LimitApplicationResult<>(
```
Do we have a tests that covers cases like `filter(limit(scan)` and `limit(filter(scan))`. In other words I think we should consider banning filter pushdown if there is already limit applied as I guess it is buggy today.

lhofhansl replied:
That's something I wanted to ask. It seems we can push down either a filter or a limit, but not both.
I debugged it and found that the framework is doing the right thing. When a filter is present it does not push the limit down (i.e. does not call applyLimit). So that works fine, and I assume the "framework" has tests for that.

kokosing replied:
Framework might have such tests but for regular tables. I think we need to add new ones for `information_schema` tables.

lhofhansl commented on presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java:
```diff
@@ -2346,6 +2346,12 @@ public void dropRole(ConnectorSession session, String role)
         return accessControlMetadata.listRoles(session);
```
ROLE_NAME, GRANTEE, and GRANTOR are defined as INFORMATION_SCHEMA.SQL_IDENTIFIER (which looks like it's not defined in the copy of the standard I have).
I think we should go by the other tables we have (ROLES, APPLICABLE_ROLES, ENABLED_ROLES), which use VARCHAR as I have done here.

The standard does not have grantor_type and grantee_type. This is an artifact Presto (and Hive really). We need those to support any connector where USERs and ROLEs are from separate domains, and we have to distinguish them.

lhofhansl commented:
Thanks for the review @kokosing .

Pushed a new change:
- added LIMIT tests. The interesting bit there was that unless we force an ordering the order is more or less random, and if we do provide an ordering the optimization cannot be used. So I'm just verifying the count except for the one case where I know internally which one row will be returned
- Cleaned up the login in SqlStandardAccessControlMetadata
- Renamed the variable names as requested.
- Fixed the logic in InformationSchemaMetadata

Didn't change:
- extra tests for InformationSchema.
- extra permission ROLE_AUTHORIZATION_DESCRIPTORS

lhofhansl commented:
Enhanced MockConnectorFactory and added role tests to TestInformationSchema.

kokosing commented on presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/file/FileHiveMetastore.java:
```diff
@@ -814,6 +814,16 @@ public synchronized void revokeRoles(Set<String> roles, Set<HivePrincipal> grant
+        ImmutableSet.Builder<RoleGrant> result = ImmutableSet.builder();
+        result.addAll(listRoleGrantsSanitized().stream()
+                .filter(grant -> grant.getRoleName().equals(role))
+                .collect(toUnmodifiableSet()));
```
Use `ImmutableSet.toImmutableSet()`, then there is no need to use `Builder`.

kokosing commented on presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/file/FileHiveMetastore.java:
```diff
@@ -826,7 +836,7 @@ public synchronized void revokeRoles(Set<String> roles, Set<HivePrincipal> grant
         result.addAll(listRoleGrantsSanitized().stream()
                 .filter(grant -> HivePrincipal.from(grant.getGrantee()).equals(principal))
-                .collect(toSet()));
+                .collect(toUnmodifiableSet()));
```
This is unrelated to the change so it should go as separate commit. 

Instead `collect` and `addAll` I would use `forEach(result::add)`

kokosing commented on presto-hive/src/main/java/io/prestosql/plugin/hive/security/SqlStandardAccessControlMetadata.java:
```diff
@@ -84,6 +88,69 @@ private static void checkRoleIsNotReserved(String role)
+        int count = 0;
+        for (String role : roles) {
+            for (RoleGrant grant : metastore.listGrantedPrincipals(role)) {
+                if (limit.isPresent() && count >= limit.getAsLong()) {
```
you should verify limit before doing a call to HMS (in outer loop), then you could simply use break.

lhofhansl replied:
Oops. That's embarrassing. :(

kokosing commented on presto-hive/src/main/java/io/prestosql/plugin/hive/security/SqlStandardAccessControlMetadata.java:
```diff
@@ -84,6 +88,69 @@ private static void checkRoleIsNotReserved(String role)
+        for (String grantee : grantees) {
+            for (PrincipalType type : new PrincipalType[]{USER, ROLE}) {
+                for (RoleGrant grant : metastore.listRoleGrants(new HivePrincipal(type, grantee))) {
+                    if (limit.isPresent() && count >= limit.getAsLong()) {
```
you should verify limit before doing a call to HMS (in the middle loop)

lhofhansl replied:
Same as above... A bit embarrassing.

kokosing commented on presto-hive/src/main/java/io/prestosql/plugin/hive/security/SqlStandardAccessControlMetadata.java:
```diff
@@ -84,6 +88,69 @@ private static void checkRoleIsNotReserved(String role)
+
+    private Set<RoleGrant> getRoleGrantsByGrantees(Set<String> grantees, OptionalLong limit)
+    {
+        ImmutableSet.Builder<RoleGrant> builder = ImmutableSet.builder();
```
s/builder/roleGrants

kokosing commented on presto-hive/src/main/java/io/prestosql/plugin/hive/security/SqlStandardAccessControlMetadata.java:
```diff
@@ -84,6 +88,69 @@ private static void checkRoleIsNotReserved(String role)
+
+    private Set<RoleGrant> getRoleGrantsByRoles(Set<String> roles, OptionalLong limit)
+    {
+        ImmutableSet.Builder<RoleGrant> builder = ImmutableSet.builder();
```
s/builder/roleGrants

kokosing commented on presto-hive/src/main/java/io/prestosql/plugin/hive/security/SqlStandardAccessControlMetadata.java:
```diff
@@ -84,6 +88,69 @@ private static void checkRoleIsNotReserved(String role)
+    @Override
+    public Set<RoleGrant> listAllRoleGrants(ConnectorSession session, Optional<Set<String>> roles, Optional<Set<String>> grantees, OptionalLong limit)
+    {
+        /*
+         * The Hive Metadata interface provides us with two ways to retrieve role grants:
+         * 1. principals by single role
+         * 2. roles by single principal
+         *
+         * We check the passed predicates to decide which path is most optimal.
+         */
+        if (grantees.isPresent()) {
+            if (roles.isPresent() && roles.get().size() <= grantees.get().size() * 2) {
+                // 2x because for the grantees we do not know the grantee type (ROLE or USER)
+                return getRoleGrantsByRoles(roles.get(), limit);
+            }
+            else {
+                return getRoleGrantsByGrantees(grantees.get(), limit);
+            }
+        }
+        else {
+            // if roles are present use those, otherwise get a list of roles from the metastore
+            return getRoleGrantsByRoles(roles.orElseGet(() -> metastore.listRoles()), limit);
+        }
```
```suggestion
        Set<String> actualRoles = roles.orElseGet(() -> metastore.listRoles());
      
        // choose more efficient path
        if (grantees.isPresent() && actualRoles.size() > grantees.get().size() * 2) {
                // 2x because we check two grantee types (ROLE or USER)
                return getRoleGrantsByGrantees(grantees.get(), limit);
        }
        return getRoleGrantsByRoles(actualRoles, limit);
        
```

lhofhansl replied:
That leads to a possibly unnecessary call to the HMS, though.

lhofhansl replied:
On the other hand it's a nice safety check to upper bind the number to RPC to no more that Roles+1. So I'll do that indeed.

kokosing commented on presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java:
```diff
@@ -4973,6 +4973,113 @@ public void testShowColumnMetadata()
+                        "('test_r_a_d1', null, null, 'user', 'USER', 'NO')");
+
+        assertQuery(
+                "SELECT * FROM information_schema.role_authorization_descriptors LIMIT 1000000000",
```
use LIMIT 5

lhofhansl replied:
I tried specifically with a very large number to make sure it does not cause any issues.

kokosing commented on presto-main/src/main/java/io/prestosql/connector/informationschema/InformationSchemaPageSource.java:
```diff
@@ -346,13 +355,35 @@ private void addRolesRecords()
+            return;
+        }
+
+        for (RoleGrant grant : metadata.listAllRoleGrants(session, catalogName, roles, grantees, limit)) {
```
Do you think it is doable to implement predicate and limit pushdown only here (instead of `SqlStandardAccessControlMetadata`)? That way it will out of the box for any other connectors or even for any access control metadata in hive. That would possibly require adding different methods to `metadata` (more granular ones).

WDYT?

lhofhansl replied:
That's what I had first, but @electrum correctly pointed out that this is a vert Hive specific.
So in this design here we just pass the predicates down, and then let the Connectors do what's best for them. (For example only HMS has this weird limit that the only APIs are getting Roles by a single Principal, or Principals by a single Role).

kokosing replied:
I see, but I am afraid that we went to far with this being too much hive specific and so it can be difficult in future to reuse it as one might need to reimplement predicate pushdown in the one use case.

> HMS has this weird limit that the only APIs are getting Roles by a single Principal, or Principals by a single Role

Could expose this in Connector metadata SPI? I would imagine that in future we could also have batch version of these methods. In case when batch version is not implemented (like in Hive) we could fallback to using methods to access roles for given principal or principals for given role.

WDYT?

lhofhansl replied:
ConnectorMetadata has now two relevant methods: (1) list all roles, and (2) list all role grants (with optional filter criteria).

I think connectors will be quite different on this front. Databases (Postgres, Oracel, MySQL, Phoenix, etc) have their own `information_schema` and the connector would just turn this into a query on *their* `information_schema` in that case we'd want to pass the predicate through. 
Others (like Hive) have limited and different APIs, and that needs some logic to be "optimal". Yet, other connectors have other way to query their role metadata.

Both methods may or may not be batched, but it's completely hidden now. The connector decides how to implement them. Note that the Hive implementation is completely hidden in the Hive Metadata, and the only part the InformationSchema needs are these two methods to be called. Everything else is automated (extracting the predicate, passing it down, etc)

If we do not want to leak Hive specifics into the InformationSchemas code I do not see how to do it fundamentally differently. If I put specific logic any higher than ConnectorMetadata we'd leak the details of Hive up.

But maybe I don't understand what you mean...
Perhaps could outline what methods you would add to ConnectorMetadata, and how would the code in InformationSchemaPageSource use them...?

Thanks!

kokosing replied:
Thanks!

kokosing commented on presto-tests/src/test/java/io/prestosql/tests/TestInformationSchemaConnector.java:
```diff
@@ -97,6 +97,17 @@ public void testLimit()
+        assertQuery("SELECT count(*) FROM test_catalog.information_schema.enabled_roles", "VALUES 50");
+        assertQuery("SELECT count(*) FROM test_catalog.information_schema.applicable_roles", "VALUES 1");
+        assertQuery("SELECT role_name FROM test_catalog.information_schema.role_authorization_descriptors WHERE grantee = 'user5'", "VALUES ('role2')");
+        assertQuery("SELECT grantee FROM test_catalog.information_schema.role_authorization_descriptors WHERE role_name = 'role2'", "VALUES ('user4'), ('user5')");
```
can you use `assertMetadataCalls` to make sure that predicate pushdown and limit works.

lhofhansl replied:
Hmm... I'd have to look inside the function calls then (i.e. I'd have to know not only that a function was called, but also what was passed). Let me think about how I'd change the CountingMockConnector to do so.

lhofhansl commented:
Thanks for the review @kokosing .

Pushed another update.
- I left the predicate pushdown where it is, since it is really only the connector that knows how to best optimize. I think this is as good as it gets, otherwise we'd be leaking Hive specific details into all other connectors.
- Also have not done the predicate pushdown counting, yet. Looking into that part now.

lhofhansl commented:
Also added predicate pushdown counting, which also verifies that LIMIT is not pushed together with the other predicates.

lhofhansl commented:
Pushed a rebase. Let's get this over the finishing line. :)

kokosing commented:
> Let's get this over the finishing line. :)

Let's solve the: https://github.com/prestosql/presto/pull/3535#discussion_r430218051

lhofhansl commented:
Thanks @kokosing . See my comment: https://github.com/prestosql/presto/pull/3535#discussion_r431424748

kokosing commented on presto-main/src/main/java/io/prestosql/connector/informationschema/InformationSchemaMetadata.java:
```diff
@@ -242,6 +251,65 @@ public static boolean isTablesEnumeratingTable(InformationSchemaTable table)
+        }
+
+        Optional<Set<String>> grantees = filterString(constraint, GRANTEE_COLUMN_HANDLE);
+        if (grantees.isPresent()) {
```
You might want to invert the condition (just to have less code nested)

kokosing commented on presto-testing/src/main/java/io/prestosql/testing/CountingMockConnector.java:
```diff
@@ -101,6 +121,19 @@ private ConnectorFactory getConnectorFactory()
+                        granteesPushedCounter.incrementAndGet();
+                    }
+                    if (limit.isPresent()) {
+                        limitPushedCounter.incrementAndGet();
```
`listRowGrantsCallsCounter`, `rolesPushedCounter`, `rolesPushedCounter` and `limitPushedCounter` should be grouped into a single bean that represents interaction with metadata call. Typically we had single counter per method, you had 4. Using bean we would maintain having single counter per method, where one counter is just complex.

lhofhansl replied:
I added beans: a Counter bean a Count bean to compare against.

lhofhansl commented:
@kokosing , pushed a new update. Let me know whether the bean approach in CountingMockConnector is what you had in mind.

lhofhansl commented:
TestThrottledAsyncQueue seems unrelated and passes locally.

kokosing commented on presto-tests/src/test/java/io/prestosql/tests/TestInformationSchemaConnector.java:
```diff
@@ -97,6 +97,48 @@ public void testLimit()
+        assertMetadataCalls(
+                "SELECT count(*) FROM (SELECT * FROM test_catalog.information_schema.role_authorization_descriptors WHERE grantee = 'user5' LIMIT 1)", "VALUES 1",
+                new MetadataCallsCount()
+                        .withRoleGrantCount(1, 0, 1, 0));
```
That is very nice. How about going one step further and having this more fluent, like 
```
new MetadataCallsCount().with(roleGrantCount(/*method calls count*/ 1).withGranteesPushedDown(1));
```
?

lhofhansl replied:
I had those methods before. It just seemed to now make more sense to compare the beans...

Maybe it was better the way I had it before?

In the end we have certain metrics we want to count for the connector; those are the atomic counters.
I think I misunderstood what you wanted... MetadataCallsCount is already a "bean", and so having the individual members for counts is fine...?

kokosing commented on presto-testing/src/main/java/io/prestosql/testing/CountingMockConnector.java:
```diff
@@ -167,6 +202,103 @@ public String toString()
+        }
+    }
+
+    public static class RoleGrantCounter
```
RoleGrantCounter?

lhofhansl replied:
Yeah... Naming is hard. :)

kokosing commented on presto-testing/src/main/java/io/prestosql/testing/CountingMockConnector.java:
```diff
@@ -167,6 +202,103 @@ public String toString()
+        private final AtomicLong listRowGrantsCallsCounter = new AtomicLong();
+        private final AtomicLong rolesPushedCounter = new AtomicLong();
+        private final AtomicLong granteesPushedCounter = new AtomicLong();
+        private final AtomicLong limitPushedCounter = new AtomicLong();
```
I would prefer to have bean to be immutable. `incrementListRoleGrants` could return new instance. Then you don't need two classes.

lhofhansl replied:
The AtomicLongs for the other counters are mutable. I think this counter would be inherently mutable.
How do you handle atomicity between calls? Have the counter bean be an atomic reference and you update the reference with each new instance?

We're going to need 2 beans, right? One for counting and one to compare against during the test.
The more I think about it, the more I think the previous code was better.

lhofhansl commented:
Pushed another update. @kokosing .
I think for the test this is as good as it gets. The counters need to be mutable (that's also why the other calls are tracked with - mutable - AtomicLongs). I consolidated the counting into a single bean that tracks the relevant counts.

kokosing commented:
Merged, thanks!

lhofhansl commented:
Thank you @kokosing . And thanks for the detailed review on this - it makes me feel really good about the quality of the Presto code base!
